Sources/TootSDK/TootClient/TootClient+ASWebAuthentication.swift

```diff
@@ -45,8 +43,6 @@
                             "There was a problem authenticating the user: no URL was returned from the first authentication step."))
                 }
 
-                authSession.presentationContextProvider = presentationContextProvider
```

Comment:
This is also interesting. From what I remember presentation context provider was always needed unless this changed or I misremember. What's the motivation behind this change?

Reply:
Lack of knowledge as to exactly what this was supporting within the original code, combine with the init of the anchor being @MainActor isolated. 

I’m definitely keen for testing and suggestions on this one !

Reply:
I also recently did an own version of this and just marked the larger function MainActor to workaround the NS/UIWindow initialiser issue.
(BTW: I find it quite weird to return an empty window object as the anchor, even though it works :-) )

Reply:
Compiled this myself and I had only error on defaultPresentationAnchor: Main actor-isolated default value in a nonisolated context.
Here is my attempt to keep the old functionality and resolve that problem. I initialize that default value if user did not provide their own provider and its not set yet. Since the whole function is isolated to MainActor it all works.

```patch
diff --git forkSrcPrefix/Sources/TootSDK/TootClient/TootClient+Account.swift forkDstPrefix/Sources/TootSDK/TootClient/TootClient+Account.swift
index 685004daee2902c3a1b59e9b388b9f36349b7138..83a643f9d9ba7c3eb4d6cf5bb8830890aba6a1c1 100644
--- forkSrcPrefix/Sources/TootSDK/TootClient/TootClient+Account.swift
+++ forkDstPrefix/Sources/TootSDK/TootClient/TootClient+Account.swift
@@ -40,7 +40,7 @@ extension TootClient {
     /// Update the user’s display and preferences.
     /// - Returns: The user’s own Account with source attribute
     public func updateCredentials(params: UpdateCredentialsParams) async throws -> Account {
-        try await requireFeature(.updateCredentials)
+        try requireFeature(.updateCredentials)
         let req = try HTTPRequestBuilder {
             $0.url = getURL(["api", "v1", "accounts", "update_credentials"])
             if self.flavour == .pixelfed {
@@ -282,8 +282,6 @@ extension TootFeature {
 
     /// Ability to edit your profile
     ///
-
-    @MainActor
     public static let updateCredentials = TootFeature(supportedFlavours: [
         .mastodon, .akkoma, .pleroma, .pixelfed, .firefish, .sharkey, .goToSocial, .catodon, .iceshrimp,
     ])
diff --git forkSrcPrefix/Sources/TootSDK/TootClient/TootClient.swift forkDstPrefix/Sources/TootSDK/TootClient/TootClient.swift
index f6a392687ca2e61152ade27da3412a6aa1d221a8..eb1d3afb2fd570b1851d6e3dd9354f009b0396d5 100644
--- forkSrcPrefix/Sources/TootSDK/TootClient/TootClient.swift
+++ forkDstPrefix/Sources/TootSDK/TootClient/TootClient.swift
@@ -52,6 +52,10 @@ public class TootClient: @unchecked Sendable {
     /// The current accessToken in use
     internal var accessToken: String?
 
+    #if canImport(AuthenticationServices) && !os(tvOS) && !os(watchOS)
+        internal var defaultPresentationAnchor: TootPresentationAnchor?
+    #endif
+
     /// Initialize a new instance of `TootClient` by optionally providing an access token for authentication.
     ///
     /// After initializing, you need to manually call ``TootClient/connect()`` in order to obtain the correct flavour of the server.
diff --git forkSrcPrefix/Sources/TootSDK/TootClient/TootClient+ASWebAuthentication.swift forkDstPrefix/Sources/TootSDK/TootClient/TootClient+ASWebAuthentication.swift
index 33f3842fe4ab4343bdeed05366f140d1d7f8aa82..ebc5113f4e6d60e6f4edd82cb51fd1eb01e26097 100644
--- forkSrcPrefix/Sources/TootSDK/TootClient/TootClient+ASWebAuthentication.swift
+++ forkDstPrefix/Sources/TootSDK/TootClient/TootClient+ASWebAuthentication.swift
@@ -21,7 +21,14 @@
             prefersEphemeralWebBrowserSession: Bool = true,
             presentationContextProvider: ASWebAuthenticationPresentationContextProviding? = nil
         ) async throws -> String {
-            
+            var presentationContextProvider = presentationContextProvider
+            if presentationContextProvider == nil {
+                if defaultPresentationAnchor == nil {
+                    defaultPresentationAnchor = TootPresentationAnchor()
+                }
+                presentationContextProvider = defaultPresentationAnchor
+            }
+
             guard
                 let callbackURLScheme: String = URL(string: callbackURI)?.scheme
             else {
@@ -43,6 +50,8 @@
                             "There was a problem authenticating the user: no URL was returned from the first authentication step."))
                 }
 
+                authSession.presentationContextProvider = presentationContextProvider
+
                 authSession.prefersEphemeralWebBrowserSession = prefersEphemeralWebBrowserSession
                 authSession.start()
             }
@@ -53,4 +62,12 @@
         }
     }
 
+    // via: https://www.andyibanez.com/posts/using-aswebauthenticationaession-swiftui/
+    public class TootPresentationAnchor: NSObject, ASWebAuthenticationPresentationContextProviding {
+
+        public func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
+            return ASPresentationAnchor()
+        }
+    }
+
 #endif
```

Reply:
Brilliant, thanks @Tunous !

Reply:
This results in this warning on my side, with a clean build etc:

![image](https://github.com/user-attachments/assets/35ba7eb5-d76c-4433-9b44-c89394db85dc)

Adding the annotation as suggested results in this:

![image](https://github.com/user-attachments/assets/be9c0799-cff5-4b85-9095-7f7a8f35023a)

Reply:
Oh, I should note I used Xcode 16 Beta 6. There are other unrelated warnings but this class they marked as MainActor.

Reply:
Thanks @Tunous I'm running Xcode 15.4 here. Good to note.

I think I'd be happy to leave this with the warning, for Xcode 15.4 @kkostov - any thoughts?

Reply:
@davidgarywood how can I get this code on my side to test it out? 

From the screenshot, I think the suggestion of Xcode is not correct, the method should be marked as `nonisolated` instead. 

To answer your question, errors in Xcode 15.x are OK, as long as they're not errors in Swift 6 or errors with strict concurrency checking turned on in Xcode 15.